[FLINK-5778] [savepoints] Add savepoint serializer with relative file path serialization

uce opened the pull request:
This adds a new savepoint version, `SavepointV2`. The corresponding `SavepointV2Serializer` is the same as our current `SavepointV1Serializer` except that `FileStateHandle` instances are serialized with their file path relative to the savepoint base path.

As an example imagine a savepoint in directory `hdfs:///path/to/savepoint-directory` with this data file:

```
hdfs:///path/to/savepoint-directory/_metadata
hdfs:///path/to/savepoint-directory/data-X
hdfs:///path/to/savepoint-directory/data-Y
```

Previously, the complete file path was stored. With this PR, we only store `data-X` for file state handles and reconstruct the complete path from the savepoint directory on restore. This enables us to move the savepoint directory around. The only requirement is that the layout within the savepoint directory does not change. I think this is a reasonable restriction.

In addition to the added tests, I've tested this manually by triggering savepoints, moving the savepoint around in the local file system as well as to HDFS and restoring from it. 

The code between `SavepointV1` and `SavepointV2` and the respective serializers is mostly shared. Therefore, I've moved the base logic out to an abstract `AbstractSavepoint` and `AbstractSavepointSerializer`.

The migration story is that you can resume old savepoints as before and all newly triggered savepoints will be V2 savepoints that serialize file state handles with their relative path. You can also resume with `1.3-SNAPSHOT` savepoint without any issues.

uce commented on flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/savepoint/AbstractSavepoint.java:
```diff
@@ -0,0 +1,94 @@
+ * A base savepoint class implementing the simple accessors of the Savepoint
+ * interface.
+ */
+abstract class AbstractSavepoint implements Savepoint {
```
This code was the same for both `SavepointV1` and `SavepointV2`. Therefore I've moved it here.

uce commented on flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/savepoint/AbstractSavepointSerializer.java:
```diff
@@ -0,0 +1,386 @@
+ * bad idea, feel free to change stuff. Right now, it's sole purpose is reducing
+ * code duplication between {@link SavepointV1Serializer} and {@link SavepointV2Serializer}.
+ */
+abstract class AbstractSavepointSerializer<T extends Savepoint> implements SavepointSerializer<T> {
```
All of this is shared between V1 and V2 savepoints. I've moved it here and allowed to overwrite the file state handle serialization behaviour. @StefanRRichter, @StephanEwen The usefulness of this depends on how many changes we expect in the future. If we expect many, I think it is better to actually duplicate the code instead of moving it to an abstract class. For now, I'm slightly leaning towards not expecting many changes and hence moved it to the abstract class. In any case, moving back with this is a very trivial refactoring if you want the code to be rather copied.

uce replied:
Agreed, let me change this.

uce commented on flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/savepoint/SavepointV2Serializer.java:
```diff
@@ -0,0 +1,109 @@
+	}
+
+	@Override
+	void serializeFileStreamStateHandle(FileStateHandle fileStateHandle, Path basePath, DataOutputStream dos) throws IOException {
```
Here is the new serialization behaviour

StefanRRichter commented on flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/savepoint/SavepointV2Serializer.java:
```diff
@@ -0,0 +1,109 @@
+			dos.writeBoolean(true);
+			dos.writeUTF(relative.toString());
+		} else {
+			dos.writeBoolean(false);
```
I think this branch is never taken in the current implementation. If it would be taken, this would silently violate the user's assumption that she can simply move around the savepoint data. I suggest to remove this branch and stick to the actual case. If we ever want to support a different behaviour in the future, we can simply do so by coming up with another version of the serializer.

uce replied:
The branch was meant to handle future changes, but you are right that this is not a good idea (for the reasons you mention). I will remove it.

uce commented:
Thanks for your review Stefan! I addressed your comments, but only then realized that the restriction to relative file state handle is actually a problem for externalized checkpoints. :-( They possibly store their meta data somewhere else and not as part of the checkpoint files. For them it's OK to not be relocatable. Still, I agree with your comment that this should be more explicitly allowed or disallowed. I gues that we should explicitly distinguish between savepoints and externalized checkpoints for the meta data serialization. I hope that Stephan's refactorings of the checkpoint stream creation logic will help here. I'll leave this PR open with your suggested changes and adjust it after Stephan's PR. Then I can either piggyback on his changes or see whether I need to add the required distinction myself.

StephanEwen commented:
Looks good in general.

One thing that I stumbled across a lot in recent works on the checkpoints / savepoints is that they all implement `equals` and `hashCode` and delegate to the task states and handles, etc. 

Do we need to define semantic equality there? It seems fragile to me, because the state handles by themselves can in general not really make a good claim about equality. The `FileStateHandle` for example fails to define equality depending on whether there is a trailing '/' on a directory path or not.

StephanEwen commented:
Another thought from the discussion with @StefanRRichter :
You refactor a lot to not have duplicate code. While this is good in general, I am wondering if we should not actually duplicate the code here, because we want the `V1` serialization code and savepoint code to be "immutable", meaning it should not be affected by changes to the `V2` code. Having a copy makes sure no accidental changes are made to the `V1` code when modifying the `V2` code.

uce commented:
I agree with the equals/hashCode question. It has been introduced for the various involved classes by different people and not in this PR, so I think it's best handled as part of a different issue. Would you mind opening an issue for it? 

Regarding code refactoring: I was also skeptical about the value of refactoring this. I'm totally fine with keeping the code duplicated. @StefanRRichter Is that OK with you as well?

StefanRRichter commented:
Very much ok with me, my vote on this was actually on keeping old versioned code duplicated and immutable.

tillrohrmann commented:
Are we gonna get this in for 1.3?

zentol commented:
What's the status of this PR @uce @tillrohrmann @StefanRRichter @StephanEwen ?

uce commented:
I had a quick chat with Stephan about this. @StefanRRichter has an idea how to properly implement this. Closing this PR and unassigning the issue.
